lib/media/autoplay.js

```diff
@@ -0,0 +1,73 @@
+/*! @license
+ * Shaka Player
+ * Copyright 2016 Google LLC
```

Comment:
This file is:

1. difficult to read
2. appears to be copied verbatim from video-dev/can-autoplay

Copying this code and calling it "Copyright Google" under an Apache license is a problem.  See the can-autoplay license:

> MIT License
> 
> Copyright (c) 2017 video-dev
> 
> Permission is hereby granted ...
> ... subject to the following conditions:
> 
> The above copyright notice and this permission notice shall be included in all
> copies or substantial portions of the Software.

So we can't incorporate this code in the way you've provided it.

Fortunately, I also find it completely unreadable, so I would prefer it to be rewritten anyway.

I would like to see:

1. The video bytes encoded in hex, in an orderly array, line-wrapped to meet our style guidelines
2. The result of the check cached in a static variable (see shaka.media.DrmEngine.DUMMY_KEY_ID defined with shaka.util.Lazy) since it will be called potentially many times
3. The method simplified and cleaned up.  I can't see any rhyme or reason to most of it (sendOutput being null initially, using `void 0` instead of null, the very existence of sendOutput instead of just calling resolve() directly, the split and callback structure between startPlayback() and video(), etc.)

Reply:
>2. appears to be copied verbatim from video-dev/can-autoplay

It is. Though I didn't pay attention to the license. My bad.


>1. The video bytes encoded in hex, in an orderly array, line-wrapped to meet our style guidelines

You mean like the content of the Buffer before converting it to Uint8? Something like what's inside this \<Buffer>:
`<Buffer 00 00 00 1c 66 74 79 70 69 73 6f 6d 00 00 02 00 69 73 6f 6d 69 73 6f 32 6d 70 34 31 00 00 00 08 66 72 65 65 00 00 02 ef 6d 64 61 74 21 10 05 20 a4 1b ...>`

For now I've cleaned up the code a little bit. I turned the null `playResult` and `sendOutput` to constants inside the Promise, but I kept the sendOutput method to call `element.remove()`.

I documented `startAds_`'s params and made it private, but I couldn't figure a way to make it work by calling `this.imaAdsManager_`. When I do that, it seems to work perfectly for all ads, except for the prerolls. I tested it with a [VAST](https://pubads.g.doubleclick.net/gampad/ads?iu=/21775744923/external/single_ad_samples&sz=640x480&cust_params=sample_ct%3Dlinear&ciu_szs=300x250%2C728x90&gdfp_req=1&output=vast&unviewed_position_start=1&env=vp&impl=s&correlator=) and a [VMAP](https://pubads.g.doubleclick.net/gampad/ads?iu=/21775744923/external/vmap_ad_samples&sz=640x480&cust_params=sample_ar%3Dpremidpost&ciu_szs=300x250&gdfp_req=1&ad_rule=1&output=vmap&unviewed_position_start=1&env=vp&impl=s&cmsid=496&vid=short_onecue&correlator=) sample tag and both failed to start on the preroll.

Reply:
As for the use of shaka.util.Lazy. Should I resolve the Promise inside `Lazy()` and use its value on the ClientSideAdsManager?:

Inside `autoplay.js`:
```
shaka.media.AutoPlay.CAN_AUTOPLAY = new shaka.util.Lazy(async () => {
  const AutoPlay = new shaka.media.AutoPlay();
  const {result} = await AutoPlay.canAutoPlay();
  return result;
});
```

And then, inside `startAds_`:
```
const canAutoPlay = await shaka.media.AutoPlay.CAN_AUTOPLAY.value();
```